### ISSUE TITLE
Combining multiple tasks

### ISSUE THREAD

**Issue by rom1504** [andrewrk/mineflayer#275]
When creating a bot, you usually want to combine many different methods and information to accomplish tasks.
Mineflayer provides unit action (digging, simple movements, building, ...) and information.

There are many ways to combine these actions to produce useful behavior : 
* the [pyramid of doom](https://github.com/andrewrk/mineflayer/blob/094a9510cb514ca56f9ce8f888aa4c6d0d7ac7d4/test/externalTests/useChests.js#L28)  (anti-)pattern
* defining case specific functions to combine functions ([this](https://github.com/andrewrk/mineflayer/blob/5675cbd0dab18bd019771327f9ab8c40b4c9536f/test/externalTests/plugins/testCommon.js#L21) for example)
* state machines (in [mineflayer-scaffold](https://github.com/andrewrk/mineflayer-scaffold/blob/master/index.js#L65))
* iterators (in [helperbot](https://github.com/Darthfett/helperbot/blob/8a9ca118d2e9e21a79091034040e304f4256f2a6/lib/plugins/stripmine.js#L197))
* meta programming (in [rbot](https://github.com/rom1504/rbot/blob/master/task/syntaxTask.js))
* a task manager (see [dutie](https://github.com/demipixel/dutie) and [dutieflayer](https://github.com/demipixel/dutieflayer) but also [duties](https://github.com/wtfaremyinitials/duties))

There are also some stuff provided by the js community that might be useful : 
* [async](https://github.com/caolan/async)
* [promises](https://github.com/kriskowal/q)

In this issue I'd like to talk about whether we can find some common ways on how to combine mineflayer tasks to get useful mineflayer. That doesn't mean these ways should be included in mineflayer, but since mineflayer is the common factor in all this, it seems a good place to talk about.

I think criteria to know whether one of these ways works or not are : 
* do you need to write a lot of code to use it ?
* is that code complicated to read ?
* how much was done with that method ? does it work to actually get useful bots ?

It might be interesting to make a table to compare all these methods (with the pros and the cons).

The discussion has also started there https://github.com/andrewrk/mineflayer/issues/182 and there https://github.com/demipixel/dutie/wiki/Example-task-scenarios.

**Comment by rom1504** [comment]
Several steps into doing something : 
* task generation : generate all the tasks you are going to do to accomplish something
* task scheduling : decide what you are going to do first
* task evaluation : actually do it using mineflayer

Most of the methods listed above mix all that together.
I think it might be interesting to separate these aspects because really they are different problems.

For task generation : it might be interesting to build some kind of graph of what you can do to get from a state A to a state B.
For example : 
* A = you have no tool ; B = you want to get some stone ; what you can do : [[get log; craft wood; craft a stick ; craft a wood pickaxe; get stone], [kill someone with stone] ; [find a stone itemStack in the world]; [get some tnt; make it explode; get the resulting stone]]

Each possible task could get a estimated time attribute in order to choose which to do.
If that graph was available, it would be easier to generate task based on what you have and what you want.

**Comment by demipixel** [comment]
That would be quite difficult to do. How would we program every possibility? How would we show that tnt is very resource unfriendly? What about getting wool? Making a farm would take longer but would be useful.

**Comment by vogonistic** [comment]
This should probably be a plugin, but it sounds like a lot of fun. You could attach providers that can answer with the amount of time to answer queries.

```
need: * axe
response from crafter: [go to chest, get things, go to bench, craft axe], time 60s
response from fetcher: [go to chest, get stone axe], time 5s
```

There could be providers for walking using different methods, inventory crafting, subcontracting to other providers and so on.

**Comment by rom1504** [comment]
Related to https://github.com/wurstmineberg/assets.wurstmineberg.de/blob/master/json/items.json.description.txt

**Comment by rom1504** [comment]
On that theme, trying to express achievement as tasks would be a good test for a task model http://minecraft.gamepedia.com/Achievement#List_of_achievements

**Comment by RealPeha** [comment]
Hi guys!
I have some idea and a small implementation of the task system for the bot. It is based on the concept of pipeline. I am not very good at this and am not sure that I implemented the correct pipe, but during development I took inspiration from [rxjs](rxjs.dev).

Here is the repository of my developments at the moment: [mineflayer-pipeline](https://github.com/RealPeha/mineflayer-pipeline)

Here you can look at the use, as an example I took a fisherman from [mineflayer native fisherman](https://github.com/PrismarineJS/mineflayer/blob/master/examples/fisherman.js): [mineflayer-pipiline fisherman](https://github.com/RealPeha/mineflayer-pipeline/blob/master/examples/fisherman.js)

I have some more ideas on adding new tasks (operators), but first I would like to find out from you how you think it makes sense to continue working on this, and also some implementation tips

Also, I did not write readme for it because I'm bad in English and this message was translated using a translator

And another question: does your project have a server in the discord?

**Comment by Slord6** [comment]
I've also been thinking about this but in the context of multiple bots working together. I think it's a use case that's worth bearing in mind. The separation of generating task/scheduling task/executing task that @rom1504 mentioned is definitely the way to go, especially in the case of cooperative bots.
For my use case I've been thinking about it in terms of a single goal, for example `Mine 1 Obsidian`, you then iterate backwards from to generate tasks that can be split between bots. So for this example, if you'll excuse the terrible diagram:

![image](https://user-images.githubusercontent.com/23644091/86823410-bef27c00-c084-11ea-9e02-040c6d6d2cba.png)

All the tasks are effectively already ordered for each bot to take the next one, provided it's sub-tasks are complete.

I think this is what was meant by "some kind of graph of what you can do to get from a state A to a state B" in the earlier comment? This seems completely feasible with mineflayer in its current state, perhaps with a couple of exceptions, eg Mining the dragon egg as a goal - not sure if you can programatically get that it's effectively a drop from the dragon?

**Comment by qrvd** [comment]
The approach of multiple bots working together sounds very interesting, it would allow for quite advanced play.
By the way, I haven't seen it mentioned here, but you may want to give Behavior Trees a look, since they're a common way to implement advanced AI and very versatile:

- [Behavior Trees - Overview](https://web.archive.org/web/20160526234139/http://aigamedev.com/open/article/bt-overview/)
- [An Introduction to Behavior Trees](https://web.archive.org/web/20160330132807/https://guineashots.com/2014/07/25/an-introduction-to-behavior-trees-part-1/)
- [Behavior Trees for AI - How they work](https://www.gamasutra.com/blogs/ChrisSimpson/20140717/221339/Behavior_trees_for_AI_How_they_work.php)

**Comment by extremeheat** [comment]
Yes, see https://github.com/PrismarineJS/mineflayer-statemachine/ for a basic finite state machine

**Comment by qrvd** [comment]
I've been trying to implement a simple Minecraft bot, and as I understand it, behavior trees are quite different to finite state machines. But both have their valid use cases

**Comment by egeres** [comment]
You guys might be interested in [STRIPS](https://github.com/primaryobjects/strips), which is a planning algorithm which could design strategies and itineraries of actions to complete a specific goal 🙌🏻
